File: backend/app/routers/conversations.py
from fastapi import APIRouter, HTTPException, Query
from app.services import conversation_service

router = APIRouter(prefix="/api/conversations", tags=["conversations"])


@router.get("")
async def get_conversations(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    start_time: str | None = Query(None),
    end_time: str | None = Query(None),
    agent_id: str | None = Query(None),
    status: str | None = Query(None, pattern="^(open|closed)$"),
):
    try:
        return await conversation_service.list_conversations(
            page=page,
            page_size=page_size,
            start_time=start_time,
            end_time=end_time,
            agent_id=agent_id,
            status=status,
        )
    except Exception as e:
        raise HTTPException(status_code=502, detail=str(e))


@router.get("/{conversation_id}")
async def get_conversation(conversation_id: str):
    try:
        return await conversation_service.get_conversation(conversation_id)
    except Exception as e:
        raise HTTPException(status_code=502, detail=str(e))


@router.get("/{conversation_id}/messages")
async def get_messages(
    conversation_id: str,
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=200),
):
    try:
        return await conversation_service.list_messages(conversation_id, page, page_size)
    except Exception as e:
        raise HTTPException(status_code=502, detail=str(e))

File: backend/src/routes/conversations.ts
import { Hono } from 'hono'
import { HTTPException } from 'hono/http-exception'
import { z } from 'zod'
import * as conversationService from '../services/conversationService.js'

export const conversationsRouter = new Hono()

const listQuerySchema = z.object({
  page: z.string().optional().transform((v) => Number(v ?? '1')),
  page_size: z.string().optional().transform((v) => Number(v ?? '20')),
  start_time: z.string().optional(),
  end_time: z.string().optional(),
  agent_id: z.string().optional(),
  status: z.enum(['open', 'closed']).optional(),
})

const messagesQuerySchema = z.object({
  page: z.string().optional().transform((v) => Number(v ?? '1')),
  page_size: z.string().optional().transform((v) => Number(v ?? '20')),
})

// GET /api/conversations
conversationsRouter.get('/', async (c) => {
  const parsed = listQuerySchema.safeParse(c.req.query())
  if (!parsed.success) {
    throw new HTTPException(400, { message: 'Invalid query parameters' })
  }

  const { page, page_size, start_time, end_time, agent_id, status } = parsed.data
  const result = await conversationService.listConversations({
    page,
    pageSize: page_size,
    startTime: start_time,
    endTime: end_time,
    agentId: agent_id,
    status,
  })

  return c.json(result)
})

// GET /api/conversations/:id
conversationsRouter.get('/:id', async (c) => {
  const id = c.req.param('id')
  const conversation = await conversationService.getConversation(id)
  return c.json(conversation)
})

// GET /api/conversations/:id/messages
conversationsRouter.get('/:id/messages', async (c) => {
  const id = c.req.param('id')
  const parsed = messagesQuerySchema.safeParse(c.req.query())
  if (!parsed.success) {
    throw new HTTPException(400, { message: 'Invalid query parameters' })
  }

  const { page, page_size } = parsed.data
  const result = await conversationService.getConversationMessages(id, page, page_size)
  return c.json(result)
})
